{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/nazif/thesis/mirscribe\n"
     ]
    }
   ],
   "source": [
    "%cd ../..\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from scripts.utils import get_nucleotide_at_position, get_nucleotides_in_interval\n",
    "from scripts.features import generate_is_mirna_column\n",
    "import os\n",
    "\n",
    "from scripts.globals import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def load_vcf_into_df(file_path):\n",
    "    \"\"\"\n",
    "    Load a VCF (Variant Call Format) file into a pandas DataFrame.\n",
    "    \n",
    "    Parameters:\n",
    "    - file_path (str): The full path to the VCF file to be loaded.\n",
    "    \n",
    "    Returns:\n",
    "    - DataFrame: A pandas DataFrame containing the VCF file data, with columns for 'chr', 'pos', 'id', 'ref', and 'alt'.\n",
    "    \n",
    "    Raises:\n",
    "    - FileNotFoundError: If the specified file does not exist.\n",
    "    - ValueError: If the file is not in the expected VCF format.\n",
    "    - Exception: For other unforeseen errors during file loading.\n",
    "    \"\"\"\n",
    "    # Ensure the file path ends with '.vcf', indicating it is a VCF file\n",
    "    if not file_path.endswith('.vcf'):\n",
    "        raise ValueError(\"The file specified does not appear to be a VCF file. Please provide a valid .vcf file.\")\n",
    "\n",
    "    try:\n",
    "        return pd.read_csv(\n",
    "            file_path,\n",
    "            sep=\"\\t\",\n",
    "            header=None,\n",
    "            names=[\"chr\", \"pos\", \"id\", \"ref\", \"alt\"],\n",
    "            comment='#',\n",
    "        )\n",
    "\n",
    "    except FileNotFoundError as e:\n",
    "        # Raise a more descriptive error if the file is not found\n",
    "        raise FileNotFoundError(f\"The file at {file_path} was not found.\") from e\n",
    "    \n",
    "    except pd.errors.EmptyDataError as e:\n",
    "        # Handle the case where the file is empty or does not contain expected data\n",
    "        raise ValueError(\n",
    "            f\"The file at {file_path} is empty or not in the expected VCF format.\"\n",
    "        ) from e\n",
    "    except Exception as e:\n",
    "        # Catch-all for other exceptions, with a generic error message\n",
    "        print(f\"An error occurred while trying to read the file at {file_path}: {e}\")\n",
    "        raise\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'PD4248a'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_path = \"data/sample_vcfs/PD4248a.vcf\"\n",
    "\n",
    "filename = file_path.split(\"/\")[-1].split(\".\")[0]\n",
    "\n",
    "\n",
    "filename"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = load_vcf_into_df(\"data/sample_vcfs/PD4248a.vcf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>728633</td>\n",
       "      <td>PD4248a_1_728633_T_G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3717853</td>\n",
       "      <td>PD4248a_1_3717853_C_A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>4609964</td>\n",
       "      <td>PD4248a_1_4609964_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12695984</td>\n",
       "      <td>PD4248a_1_12695984_C_G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15821838</td>\n",
       "      <td>PD4248a_1_15821838_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr       pos                      id ref alt\n",
       "0   1    728633    PD4248a_1_728633_T_G   T   G\n",
       "1   1   3717853   PD4248a_1_3717853_C_A   C   A\n",
       "2   1   4609964   PD4248a_1_4609964_G_A   G   A\n",
       "3   1  12695984  PD4248a_1_12695984_C_G   C   G\n",
       "4   1  15821838  PD4248a_1_15821838_G_A   G   A"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def augment_id(df):\n",
    "    \"\"\"Augment the 'id' column with additional information.\"\"\"\n",
    "    df['id'] += '_' + df['chr'] + '_' + df['pos'].astype(str) + '_' + df['ref'] + '_' + df['alt']\n",
    "\n",
    "\n",
    "augment_id(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>alt_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>728633</td>\n",
       "      <td>PD4248a_1_728633_T_G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3717853</td>\n",
       "      <td>PD4248a_1_3717853_C_A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>4609964</td>\n",
       "      <td>PD4248a_1_4609964_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12695984</td>\n",
       "      <td>PD4248a_1_12695984_C_G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15821838</td>\n",
       "      <td>PD4248a_1_15821838_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr       pos                      id ref alt  ref_len  alt_len\n",
       "0   1    728633    PD4248a_1_728633_T_G   T   G        1        1\n",
       "1   1   3717853   PD4248a_1_3717853_C_A   C   A        1        1\n",
       "2   1   4609964   PD4248a_1_4609964_G_A   G   A        1        1\n",
       "3   1  12695984  PD4248a_1_12695984_C_G   C   G        1        1\n",
       "4   1  15821838  PD4248a_1_15821838_G_A   G   A        1        1"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def compute_allele_lengths(df):\n",
    "    \"\"\"Compute lengths of reference and alternate alleles.\"\"\"\n",
    "    df[\"ref_len\"] = df[\"ref\"].apply(len)\n",
    "    df[\"alt_len\"] = df[\"alt\"].apply(len)\n",
    "\n",
    "compute_allele_lengths(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>alt_len</th>\n",
       "      <th>fetched_nucleotides</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>728633</td>\n",
       "      <td>PD4248a_1_728633_T_G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3717853</td>\n",
       "      <td>PD4248a_1_3717853_C_A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>4609964</td>\n",
       "      <td>PD4248a_1_4609964_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12695984</td>\n",
       "      <td>PD4248a_1_12695984_C_G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15821838</td>\n",
       "      <td>PD4248a_1_15821838_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr       pos                      id ref alt  ref_len  alt_len  \\\n",
       "0   1    728633    PD4248a_1_728633_T_G   T   G        1        1   \n",
       "1   1   3717853   PD4248a_1_3717853_C_A   C   A        1        1   \n",
       "2   1   4609964   PD4248a_1_4609964_G_A   G   A        1        1   \n",
       "3   1  12695984  PD4248a_1_12695984_C_G   C   G        1        1   \n",
       "4   1  15821838  PD4248a_1_15821838_G_A   G   A        1        1   \n",
       "\n",
       "  fetched_nucleotides  \n",
       "0                   T  \n",
       "1                   C  \n",
       "2                   G  \n",
       "3                   C  \n",
       "4                   G  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def fetch_nucleotides(df):\n",
    "    # Function to fetch nucleotides based on the length of the affected sequences\n",
    "    # if affected sequences are longer than 1, fetch nucleotides at the specified interval\n",
    "    df.loc[df['ref_len'] > 1, 'fetched_nucleotides'] = df.apply(\n",
    "        lambda x: get_nucleotides_in_interval(x['chr'], x['pos'], x[\"pos\"]+x[\"ref_len\"]-1), axis=1\n",
    "    )\n",
    "    # if affected sequences are equal to 1, fetch nucleotides at the specified pos\n",
    "    df.loc[df['ref_len'] == 1, 'fetched_nucleotides'] = df.apply(\n",
    "        lambda x: get_nucleotide_at_position(x['chr'], x['pos']), axis=1\n",
    "    )\n",
    "    # same but for 0 length\n",
    "    df.loc[df['ref_len'] == 0, 'fetched_nucleotides'] = \"\"\n",
    "    \n",
    "\n",
    "fetch_nucleotides(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>id</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>ref_len</th>\n",
       "      <th>alt_len</th>\n",
       "      <th>fetched_nucleotides</th>\n",
       "      <th>is_nucleotides_same</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>728633</td>\n",
       "      <td>PD4248a_1_728633_T_G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>T</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>3717853</td>\n",
       "      <td>PD4248a_1_3717853_C_A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>4609964</td>\n",
       "      <td>PD4248a_1_4609964_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>12695984</td>\n",
       "      <td>PD4248a_1_12695984_C_G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>15821838</td>\n",
       "      <td>PD4248a_1_15821838_G_A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chr       pos                      id ref alt  ref_len  alt_len  \\\n",
       "0   1    728633    PD4248a_1_728633_T_G   T   G        1        1   \n",
       "1   1   3717853   PD4248a_1_3717853_C_A   C   A        1        1   \n",
       "2   1   4609964   PD4248a_1_4609964_G_A   G   A        1        1   \n",
       "3   1  12695984  PD4248a_1_12695984_C_G   C   G        1        1   \n",
       "4   1  15821838  PD4248a_1_15821838_G_A   G   A        1        1   \n",
       "\n",
       "  fetched_nucleotides  is_nucleotides_same  \n",
       "0                   T                 True  \n",
       "1                   C                 True  \n",
       "2                   G                 True  \n",
       "3                   C                 True  \n",
       "4                   G                 True  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def compare_nucleotides(df):\n",
    "    # Compare the fetched nucleotides with the reference nucleotides\n",
    "    df[\"is_nucleotides_same\"] = df[\"fetched_nucleotides\"] == df[\"ref\"]\n",
    "\n",
    "compare_nucleotides(df)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_sequences(df):\n",
    "    \"\"\"\n",
    "    Prepares reference indices, extended sequences, wild type, and mutated sequences for a given DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    - df (pd.DataFrame): The DataFrame containing the genetic information.\n",
    "    - get_nucleotides_in_interval (callable): A function that takes chromosome ('chr'), start, and end positions as arguments and returns the nucleotide sequence.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: The original DataFrame with added columns for reference start/end, upstream/downstream sequences, wild type sequence, and mutated sequence.\n",
    "    \"\"\"\n",
    "    # Preparing reference indices\n",
    "    df[\"ref_start\"] = df[\"pos\"]\n",
    "    df[\"ref_end\"] = df[\"pos\"] + df[\"ref_len\"]\n",
    "\n",
    "    # Preparing extended sequences\n",
    "    df['upstream_sequence'] = df.apply(lambda row: get_nucleotides_in_interval(row['chr'], row['ref_start'] - 30, row[\"ref_start\"] - 1), axis=1)\n",
    "    df['downstream_sequence'] = df.apply(lambda row: get_nucleotides_in_interval(row['chr'], row['ref_end'], row[\"ref_end\"] + 29), axis=1)\n",
    "\n",
    "    # Preparing wild type sequence\n",
    "    df['sequence'] = df[\"upstream_sequence\"] + df[\"ref\"] + df[\"downstream_sequence\"]\n",
    "\n",
    "    # Preparing mutated sequence\n",
    "    df['mutated_sequence'] = df[\"upstream_sequence\"] + df[\"alt\"] + df[\"downstream_sequence\"]\n",
    "\n",
    "\n",
    "prepare_sequences(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /home/nazif/thesis/data/pyensembl/GRCh37/ensembl75/Homo_sapiens.GRCh37.75.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "def import_pyensembl(grch):\n",
    "    if grch not in [37, 38]:\n",
    "        raise ValueError(\"grch must be either 37 or 38\")\n",
    "\n",
    "    ens_release = 75 if grch == 37 else 109\n",
    "    \n",
    "    import os\n",
    "\n",
    "    from pyensembl import EnsemblRelease\n",
    "    os.environ['PYENSEMBL_CACHE_DIR'] = PYENSEMBL_CACHE_DIR\n",
    "    assembly = EnsemblRelease(ens_release)\n",
    "    assembly.download()\n",
    "    assembly.index()\n",
    "    \n",
    "    return assembly\n",
    "\n",
    "grch37 = import_pyensembl(37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = generate_is_mirna_column(df, grch=37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_transcript_id_and_gene_name_columns(df, assembly):\n",
    "    \n",
    "    df['transcript_id'] = df.apply(lambda x: assembly.transcript_ids_at_locus(x['chr'], x['pos']), axis=1)\n",
    "    # df[\"transcript_name\"] = df.apply(lambda x: assembly.transcript_names_at_locus(x['chr'], x['pos']), axis=1)\n",
    "    df[\"gene_name\"] = df.apply(lambda x: assembly.gene_names_at_locus(x['chr'], x['pos']), axis=1)\n",
    "    \n",
    "\n",
    "generate_transcript_id_and_gene_name_columns(df, grch37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_and_save(df, file_name, save_path=\"results\"):\n",
    "    \"\"\"Classifies mutations, saves to files, and logs the output.\"\"\"\n",
    "\n",
    "    case_1 = df[df.is_mirna == 0][[\"id\", \"sequence\", \"mutated_sequence\"]]\n",
    "    case_2 = df[df.is_mirna == 1][[\"id\", \"sequence\", \"mutated_sequence\"]]\n",
    "\n",
    "    # Create the directory if it does not exist\n",
    "    os.makedirs(save_path, exist_ok=True)\n",
    "\n",
    "    case_1.to_csv(f\"{save_path}/{file_name}_case_1.csv\", index=False)\n",
    "    if not case_2.empty:\n",
    "        print(f\"{len(case_2)} case 2 mutations were found on {file_name}.\")\n",
    "        case_2.to_csv(f\"{save_path}/{file_name}_case_2.csv\", index=False)\n",
    "    else:\n",
    "        print(f\"No case 2 mutations were found on {file_name}.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No case 2 mutations were found on PD4248a.\n"
     ]
    }
   ],
   "source": [
    "classify_and_save(df, filename)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mirscribe-dLLZ3oWm-py3.12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
